### PR TITLE
buildPythonPackage: use wheels internally

### LINKED ISSUE #329: Python: switch to wheel format
It has been accepted as *the* binary format for Python.

See:

- https://pypi.python.org/pypi/wheel/0.15.0
- http://www.python.org/dev/peps/pep-0427/

Basically we switch from *bdist_egg* to *bdist_wheel*.

### PR DESCRIPTION
**Pull request by domenkozar**
Wheels [PEP427](https://www.python.org/dev/peps/pep-0427/) are the new python packaging binary format. We should use wheels as that's where currently python packaging development is focused on and it's considered best practice toolchain. Wheels deprecate eggs.
### Changes
- pypy installs packages to $out/site-packages by default and we now respect that default
- introduce `bootstrapped-pip` package that includes a hand crafted pip+wheel packages needed in `buildPythonPackage`
- remove our ugly hack with inserting `setuptools` before `distutils` and use exact code that pip uses
- remove support for `.pth` recursive loader. This didn't work for last 2 years, it has a major startup time impact on binaries and it's not needed since all python interpreters now use `setup-hook` to set `$PYTHONPATH`
- deprecated `pythonPath` argument to `buildPythonPackage` since it's essentially the same as `propagatedBuildInputs`
- replace setup.py build/install commands with wheel equivalent
- removed `setupPyBuildFlags` since it no longer applies
- bump pypy to 4.0
-  script that raises a conflict if two or more versions of the same package are in $PYTHONPATH
### TODO
- [x] submit a PR upstream for the `pip install --prefix=` patch https://github.com/pypa/pip/pull/3252
- [x] support `setupPyBuildFlags`
- [x] test if [data_files are not installed at correct directory](https://bitbucket.org/pypa/wheel/issues/120/bdist_wheel-creates-a-wheel-that-installs) 
- [x] test pypy support
- [x] [make sure hydra builds all packages successfully](https://headcounter.org/hydra/jobset/iElectric/python-wip)
- [x] update documentation
- [x] fix `pip install -e --prefix`
### NOTES
- setuptools namespaces support is still broken, but that's not a regression
- I had to disable zc.recipe.egg due to a bug in setuptools (reported upstream)

cc @garbas @offlinehacker @FRidh @cillianderoiste @madjar @johbo @chaoflow @datakurre @adnelson 

Fixes https://github.com/NixOS/nixpkgs/issues/329

### CODE REVIEW

**Comment by johbo** [comment]
Had a first quick look, from just reading it looks good to me. Did not manage to spot if both patch files are needed.
I'll try to test this soon on our packages, to find out if the change is smooth.

**Comment by domenkozar** [comment]
I've written a python script for checking conflicts in closure, does following output/message make sense:

```
Found duplicated packages for dependency 'requests': 
  requests 2.8.1 (/nix/store/56l5v7n13in432m6gr8qhlxzs1pbmwyd-python2.7-requests-2.8.1/lib/python2.7/site-packages)
  requests 1.2.3 (/nix/store/n8mzyqm706mf9n5pxpyr9z1nqxscgm49-python2.7-requests-1.2.3/lib/python2.7/site-packages)

Traceback (most recent call last):
  File "do_conflict.py", line 24, in <module>
    'Package duplicates found in closure, see above. Usually '
ValueError: Package duplicates found in closure, see above. Usually happens if two packages depend on different version of the same dependency.
```

**Comment by johbo** [comment]
One small suggestion for the message to add "this" after "Usually":

```
see above. Usually this happens if two
```

File: pkgs/top-level/python-packages.nix [hunk 1]
```diff
@@ -126,6 +126,8 @@ let
     mpi = pkgs.openmpi;
   };
 
+  bootstrapped-pip = callPackage ../development/python-modules/bootstrapped-pip { };
```
Comment (garbas) [inline pkgs/top-level/python-packages.nix]: is this actually needed? i see bootstrapped-pip as an implementation detail of ``development/python-modules/generic/``

Comment (domenkozar) [inline pkgs/top-level/python-packages.nix]: It's not needed, but I don't see a reason to hide it from package set.

Comment (garbas) [inline pkgs/top-level/python-packages.nix]: possible reason could be to not confuse newbies which pip to use.
another would be that our all-packages.nix is growing and if we can take something from it we should.

Comment (johbo) [inline pkgs/top-level/python-packages.nix]: Maybe it helps to ask if we have a use case for having the bootstrap pip available there? If not, we can hide it until we have one.

Comment (domenkozar) [inline pkgs/top-level/python-packages.nix]: I'll remove it, I don't care about it too much.

Comment (domenkozar) [inline pkgs/top-level/python-packages.nix]: pushed a fix

**Comment by domenkozar** [comment]
About 2k packages fail to build, this is going to take a while :) https://headcounter.org/hydra/jobset/iElectric/python-wip

**Comment by FRidh** [comment]
This seems to occur quite often:

Found duplicated packages in closure for dependency 'pip':
  pip 7.1.2 (/nix/store/disif394n3vhf4fylnd187347hgdbqlj-python-3.4.3-bootstrapped-pip-7.1.2/lib/python3.4/site-packages)
  pip 6.0.8 (/nix/store/yryhm9ca2jd8h6iyqvw7xclbisvp7wkz-python3-3.4.3/lib/python3.4/site-packages)

**Comment by domenkozar** [comment]
@FRidh can you give a simple example package?

**Comment by FRidh** [comment]
lxml: https://headcounter.org/hydra/build/785937/nixlog/1/raw

Another error I've seen quite often, is that packages fail to successfully
import numpy.
Actually, it seems to occur even more than the one above:
pysoundfile: https://headcounter.org/hydra/build/779197/nixlog/2/raw
matplotlib: https://headcounter.org/hydra/build/781259/nixlog/1/raw
bottleneck: https://headcounter.org/hydra/build/785443/nixlog/1/raw
Basically any package depending on numpy has this.

**Comment by domenkozar** [comment]
Yeah I know about numpy, not sure what''s wrong (it seems like modules don't compile correctly due to flags)

**Comment by domenkozar** [comment]
OK, I've pushed next batch of fixes, let's see (numpy is still broken).

**Comment by domenkozar** [comment]
There are [around 200 failures](https://headcounter.org/hydra/jobset/iElectric/python-wip#tabs-evaluations), mostly numpy (which I'll try to fix asap). I consider this ready for review now. I'll be collecting reviews until 26th November and then merge this refactoring if no obstacles are in the way.

**Comment by FRidh** [comment]
Only 40 failing.  Nice work!
- scikitlearn and librosa are due to #10983 
- inginious has network issues, so unrelated
- h5py fails because hdf5 seems to be broken
- spyder uses distutils and is in the process of supporting setuptools.

**Comment by domenkozar** [comment]
@FRidh I've actually managed to compile scikitlearn locally, I'll push that in a few min

**Comment by domenkozar** [comment]
I've fixed most of the failures on `python*Packages` attrset, I've now expanded the job to test all uses of `buildPythonPackage` in nixpkgs.

**Comment by domenkozar** [comment]
according to https://headcounter.org/hydra/jobset/iElectric/python-wip only scikitlearn is failing which is noted in https://github.com/NixOS/nixpkgs/issues/10983.

**Comment by domenkozar** [comment]
I'll write update documentation then I'll plan to merge it on Wednesday.

**Comment by garbas** [comment]
@iElectric thank you for the work. i will test my setups over the weekend and
will let you know the results.

:beers:

**Comment by domenkozar** [comment]
Also updated docs, all set now.

**Comment by domenkozar** [comment]
@garbas any results? Just a reminder, I plan to merge this in 2 days.

File: doc/language-support.xml [hunk 2]
```diff
@@ -196,12 +196,12 @@ you need it.</para>
 
 <para>
   Currently supported interpreters are <varname>python26</varname>, <varname>python27</varname>,
-  <varname>python32</varname>, <varname>python33</varname>, <varname>python34</varname>
+  <varname>python33</varname>, <varname>python34</varname>, <varname>python35</varname>
```
Comment (spinus) [inline doc/language-support.xml]: This is a good oportunity for me to ask question about how backward compatibility works in nix. Are you abandoning python3.2 in current unstable and that's why it's ok?

Comment (domenkozar) [inline doc/language-support.xml]: We follow upstream which will stop receiving security updates in Feb 2016 https://www.python.org/dev/peps/pep-0392/#lifespan

You can still use 3.2 packages, we just don't officially maintain them and provide binaries for it. 

3.2 was a big pain to maintain so the reason people were still doing that was Ubuntu TLS 12.04 that got replaced with 14.04.

**Comment by garbas** [comment]
i gave this PR a try and for some reason it does not work with ``nix-shell``. here is what i get:

``Installing collected packages: XXX
  Running setup.py develop for XXX
    Complete output from command /nix/store/rfcsnszaw2jlkqdgg65h3phss8xa0mlp-python-2.7.10/bin/python2.7 -c "import setuptools, tokenize; __file__='/home/rok/dev/clients/XXX/backend/setup.py'; exec(compile(getattr(tokenize, 'open', open)(__file__).read().replace('\r\n', '\n'), __file__, 'exec'))" develop --no-deps:
    running develop
    error: can't create or remove files in install directory
    
    The following error occurred while trying to add or remove files in the
    installation directory:
    
        [Errno 30] Read-only file system: '/nix/store/rfcsnszaw2jlkqdgg65h3phss8xa0mlp-python-2.7.10/lib/python2.7/site-packages/test-easy-install-18363.write-test'
    
    The installation directory you specified (via --install-dir, --prefix, or
    the distutils default setting) was:
    
        /nix/store/rfcsnszaw2jlkqdgg65h3phss8xa0mlp-python-2.7.10/lib/python2.7/site-packages/
    
    Perhaps your account does not have write access to this directory?  If the
    installation directory is a system-owned directory, you may need to sign in
    as the administrator or "root" account.  If you do not have administrative
    access to this machine, you may wish to choose a different installation
    directory, preferably one that is listed in your PYTHONPATH environment
    variable.
    
    For information on other options, you may wish to consult the
    documentation at:
    
      https://pythonhosted.org/setuptools/easy_install.html
    
    Please make the appropriate changes for your system and try again.
``
I guess we need to solve this before we merge this PR.

**Comment by domenkozar** [comment]
@garbas good catch, something to fix upstream.

File: pkgs/development/python-modules/generic/default.nix [hunk 3]
```diff
@@ -79,86 +68,47 @@ python.stdenv.mkDerivation (builtins.removeAttrs attrs ["disabled"] // {
...
-
-      runHook postCheck
-  '';
-
   buildPhase = attrs.buildPhase or ''
     runHook preBuild
-
-    ${python}/bin/${python.executable} setup.py build ${lib.concatStringsSep " " setupPyBuildFlags}
-
+    cp ${setuppy} nix_run_setup.py
```
Comment (adnelson) [inline pkgs/development/python-modules/generic/default.nix]: Why is this `cp` happening? This is a little strange and confusing. Why not just do

```
${python.interpreter} ${./run_setup.py} ${lib.optionalString ....
```

?

Comment (domenkozar) [inline pkgs/development/python-modules/generic/default.nix]: It needs to happen in source directory, I didn't want to hack around that since I'm sure a lot of `setup.py` files assume they're at the right place

File: pkgs/development/python-modules/generic/run_setup.py [hunk 4]
```diff
@@ -0,0 +1,6 @@
+import setuptools
```
Comment (adnelson) [inline pkgs/development/python-modules/generic/run_setup.py]: This file needs some documentation. It's super weird that we're doing this.

File: pkgs/development/python-modules/generic/default.nix [hunk 5]
```diff
@@ -59,19 +49,18 @@ if disabled
 then throw "${name} not supported for interpreter ${python.executable}"
 else
 
+let
+  setuppy = ./run_setup.py;
```
Comment (adnelson) [inline pkgs/development/python-modules/generic/default.nix]: I'm still not sure what this is doing, but I've got to say I'm not a fan. This is going to confuse the heck out of developers trying to grok what `buildPythonPackage` is doing, and make error messages a lot more confusing. At the very least it needs some comments explaining the thinking.

File: pkgs/development/python-modules/generic/run_setup.py (hunk 4 above)
Comment (domenkozar) [inline pkgs/development/python-modules/generic/run_setup.py]: I agree - it's FAR from ideal, but it's the best method out there. I'll add docs about this approach.

**Comment by domenkozar** [comment]
Fixed.

File: pkgs/development/python-modules/generic/default.nix (hunk 3 above)
Comment (adnelson) [inline pkgs/development/python-modules/generic/default.nix]: I'm not sure why it would need to happen in the source directory; if it's referring to `'setup.py'` then it should look in the `cwd` of the process, not relative to the path the file is located in. If it's really necessary to put it in the same folder, there should be comments explaining why.

File: pkgs/development/python-modules/generic/catch_conflicts.py [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+            # some exceptions inside buildPythonPackage
+            if req.project_name in ['setuptools', 'pip']:
+                continue
+            packages[req.project_name].append(req)
+
+
+for name, duplicates in packages.items():
+    if len(duplicates) > 1:
+        do_abort = True
+        print("Found duplicated packages in closure for dependency '{}': ".format(name))
```
Comment (adnelson) [inline pkgs/development/python-modules/generic/catch_conflicts.py]: This might not be an error in all cases. I think a developer should be able to downgrade this to a warning only with a flag in `buildPythonPackage`.

File: pkgs/development/python-modules/generic/default.nix [hunk 7]
```diff
@@ -166,7 +116,8 @@ python.stdenv.mkDerivation (builtins.removeAttrs attrs ["disabled"] // {
        tmp_path=$(mktemp -d)
```
Comment (adnelson) [inline pkgs/development/python-modules/generic/default.nix]: While we're at it, could we add logic that only performs these steps if the setup.py is for the package we're trying to install? The current method creates very strange behavior if you run `nix-shell` for one package in another package's folder. Something along the lines of:

```bash
# Check if setup.py exists
if test -e setup.py; then
  # Check that the name of the package we're building matches the one in setup.py
  if echo ${name} | grep -q $(${python.interpreter} setup.py --name); then
    ...
  fi
fi
```

File: pkgs/development/python-modules/numpy-scipy-support.nix [hunk 8]
```diff
@@ -16,30 +16,9 @@
   # .test() function, which will run the test suite.
```
Comment (adnelson) [inline pkgs/development/python-modules/numpy-scipy-support.nix]: This comment is now out-of-date because we're not installing anything.

File: pkgs/development/python-modules/generic/default.nix [hunk 9]
```diff
@@ -12,28 +12,18 @@
...
 # enable tests by default
 , doCheck ? true
 
-# List of packages that should be added to the PYTHONPATH
-# environment variable in programs built by this function.  Packages
-# in the standard `propagatedBuildInputs' variable are also added.
-# The difference is that `pythonPath' is not propagated to the user
-# environment.  This is preferrable for programs because it doesn't
-# pollute the user environment.
+# DEPRECATED: use propagatedBuildInputs
```
Comment (adnelson) [inline pkgs/development/python-modules/generic/default.nix]: `pythonPath` and `propagatedBuildInputs` serve different roles, however. The goal of `pythonPath` is to register a package as a runtime dependency, but hide its `PYTHONPATH` from being propagated to downstream packages. This is a very desirable thing to have IMO. Why do we not want to support it?

File: pkgs/development/python-modules/generic/default.nix [hunk 10]
```diff
@@ -79,86 +68,47 @@ python.stdenv.mkDerivation (builtins.removeAttrs attrs ["disabled"] // {
...
-    rm -f "$out/lib/${python.libPrefix}"/site-packages/site.py*
+    pushd dist
+    ${bootstrapped-pip}/bin/pip install *.whl --no-index --prefix=$out --no-cache
+    popd
 
     runHook postInstall
   '';
 
+  doInstallCheck = doCheck;
+  doCheck = false;
```
Comment (adnelson) [inline pkgs/development/python-modules/generic/default.nix]: ??? explain?

File: pkgs/development/python-modules/generic/catch_conflicts.py (hunk 6 above)
Comment (domenkozar) [inline pkgs/development/python-modules/generic/catch_conflicts.py]: Can you give me an example? Python doesn't allow multi version installation, so this shouldn't be ever needed.

File: pkgs/development/python-modules/generic/default.nix (hunk 7 above)
Comment (domenkozar) [inline pkgs/development/python-modules/generic/default.nix]: This is out of scope for this refactoring. I'm introducing wheels, not trying to fix everything.

File: pkgs/development/python-modules/generic/default.nix (hunk 10 above)
Comment (domenkozar) [inline pkgs/development/python-modules/generic/default.nix]: We're enabling `doInstallCheck` which runs `checkPhase` after `installPhase`

Comment (globin) [inline pkgs/development/python-modules/generic/default.nix]: Could you explain that in a comment? On first glance it looks like it is disabling `doCheck` and then setting `doInstallCheck` to that same `false`

File: pkgs/development/python-modules/generic/catch_conflicts.py (hunk 6 above)
Comment (adnelson) [inline pkgs/development/python-modules/generic/catch_conflicts.py]: The one I encountered is `ansible`. You usually just want to use it as an executable, in which case you don't want all of its dependencies' pythonpaths showing up in your environment (the paths are just preserved in the wrapped executable).

File: pkgs/development/python-modules/generic/default.nix (hunk 7 above)
Comment (adnelson) [inline pkgs/development/python-modules/generic/default.nix]: Fair enough, yeah that's a good point

File: pkgs/development/python-modules/generic/catch_conflicts.py (hunk 6 above)
Comment (domenkozar) [inline pkgs/development/python-modules/generic/catch_conflicts.py]: You mean that's a reason to keep `pythonPath`? I'd rather have `isExecutable` flag then to control such things. In general you want python dependencies to propagate if they're usage at runtime.

Comment (adnelson) [inline pkgs/development/python-modules/generic/catch_conflicts.py]: I'd be fine with an `isExecutable` flag (or perhaps to be more explicit, a `dontPropagatePythonPath` flag)!

File: pkgs/development/python-modules/generic/default.nix (hunk 10 above)
Comment (layus) [inline pkgs/development/python-modules/generic/default.nix]: As long as the set is not recursive ('rec' keyword) this works. But it is really calling for problems in the future and for readability.
Maybe put doCheck in a temporary variable or move these two lines in another merged by itself ?

```nix
{ /*...*/ } // {
  doInstallCheck = doCheck;
} // {
  doCheck = false;
  #...
}
```

Comment (adnelson) [inline pkgs/development/python-modules/generic/default.nix]: It seems like there are more straightforward solutions  to this. For example, one really simple one is to just set `checkPhase = "true"` and `doInstallCheck = doCheck`.

Comment (domenkozar) [inline pkgs/development/python-modules/generic/default.nix]: I agree this is ugly. I'll find a way to make it sane.

**Comment by domenkozar** [comment]
Thank you all for feedback, I'll try to document things better today and merge this.

File: pkgs/development/python-modules/generic/catch_conflicts.py (hunk 6 above)
Comment (domenkozar) [inline pkgs/development/python-modules/generic/catch_conflicts.py]: Yeah. Let's discuss this after we merge this. `pythonPath` still works, it's just deprecated. Whatever we decide, `pythonPath` is a really bad name.

File: pkgs/development/python-modules/generic/default.nix (hunk 9 above)
Comment (domenkozar) [inline pkgs/development/python-modules/generic/default.nix]: I've answered at the question above

**Comment by domenkozar** [comment]
I've merged this, please open issues for further improvements. Thank you for feedback!

**Comment by garbas** [comment]
awesome! more :beers: for @iElectric

File: pkgs/development/python-modules/generic/run_setup.py (hunk 4 above)
Comment (adnelson) [inline pkgs/development/python-modules/generic/run_setup.py]: @domenkozar What happened to the docs for this? I'm a little disappointed that this got merged in this state.

Comment (domenkozar) [inline pkgs/development/python-modules/generic/run_setup.py]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/python-modules/generic/default.nix#L53

**Comment by datakurre** [comment]
@domenkozar I'm running into issues with cyclic import I did not have before internal wheel usage.

For example, with https://gist.github.com/datakurre/60155accd3c9be123f10#file-default-nix

I get

```
Copying plone.app.controlpanel.egg-info to build/bdist.macosx-10.7-x86_64/wheel/plone.app.controlpanel-3.0.3-py2.7.egg-info
Installing build/bdist.macosx-10.7-x86_64/wheel/plone.app.controlpanel-3.0.3-py2.7-nspkg.pth
running install_scripts
creating build/bdist.macosx-10.7-x86_64/wheel/plone.app.controlpanel-3.0.3.dist-info/WHEEL
installing
/private/var/folders/y6/2zdynm9x3nv6sq5p6780dwdc0000gn/T/nix-build-python2.7-plone.app.controlpanel-3.0.3.drv-0/plone.app.controlpanel-3.0.3/dist /private/var/folders/y6/2zdynm9x3nv6sq5p6780dwdc0000gn/T/nix-build-python2.7-plone.app.controlpanel-3.0.3.drv-0/plone.app.controlpanel-3.0.3
Ignoring indexes: https://pypi.python.org/simple
Processing ./plone.app.controlpanel-3.0.3-py2-none-any.whl
Collecting Products.CMFPlone (from plone.app.controlpanel==3.0.3)
  Could not find a version that satisfies the requirement Products.CMFPlone (from plone.app.controlpanel==3.0.3) (from versions: )
No matching distribution found for Products.CMFPlone (from plone.app.controlpanel==3.0.3)
builder for ‘/nix/store/dckn5ij1wh8nx2hin0rm9b70sam05dgw-python2.7-plone.app.controlpanel-3.0.3.drv’ failed with exit code 1
```

because I've removed CMFPlone from plone.app.controlpanel propagated build inputs to avoid cyclic dependencies.

Any tips, how should I fix cyclic dependencies now?

**Comment by datakurre** [comment]
I resolved cyclic dependencies by sedding conflicting packages off from setup.py

**Comment by datakurre** [comment]
I'm still struggling with namespace packages after this change to use wheels internally.

In short, why does namespaces work when package with namespace packages (e.g. zope.deprecation) is in PYTHONPATH, but fail when the same path is added into sys.path? @domenkozar @cillianderoiste 

Particularly, all zc.buildout generated scripts with /nix/store paths are broken because of this.
